src/common/utils/with-error-logging/with-error-logging.injectable.ts

```
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+import { getInjectable } from "@ogre-tools/injectable";
+import loggerInjectable from "../../logger.injectable";
+
+export type WithErrorLoggingFor = (
+  getErrorMessage: (error: Error) => string
+) => <T extends (...args: any[]) => any>(
+  toBeDecorated: T
+) => (...args: Parameters<T>) => ReturnType<T>;
+
+const withErrorLoggingInjectable = getInjectable({
+  id: "with-error-logging",
+
+  instantiate: (di): WithErrorLoggingFor => {
+    const logger = di.inject(loggerInjectable);
+
+    return (getErrorMessage) =>
+      (toBeDecorated) =>
+        (...args) => {
+          try {
+            const returnValue = toBeDecorated(...args);
+
+            if (isPromise(returnValue)) {
+              returnValue.catch((e: Error) => {
+                logger.error(getErrorMessage(e as Error), e);
+              });
+            }
+
+            return returnValue;
+          } catch (e) {
+            logger.error(getErrorMessage(e as Error), e);
+            throw e;
```

If we log the error, why do we also want to throw it? Then someone up the chain as well would have to deal with it.

---

Especially since we don't do that in the `async` case.

---

Tests will describe the exact behaviour and make sure it works like that always.

We are re-throwing because this higher order function has one responsibility, which is to log the error. This is related to the fact that there is lot to improve in our code that handles errors. At the moment the error handling is part of the operational code, which it shouldn't be. It could be organized in a way that error (fatal) is catched from higher level (carefully chosen place), which makes sure that application remains usable. 

`async` case actually re-throws. Tests verify that.

---

Ah I was mistaken about what `.catch` does to a promise. I thought it swallowed the error (rejected) state, but it does not.